### PR TITLE
WIP: embeding commands poc

### PR DESCRIPTION
**Pull request by ankithans**
## Issues
- embeding multiple plugins in one host CLI (multiple yaml files)
- Commands are interfering each other when multiple commands are present in yaml
- we are assuming that user's system has CLI installed which is to be embedded

## steps to run
1. Run `make`
2. `./embed`

taken help from: https://github.com/craicoverflow/socket

#26 #25

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -1,202 +1,202 @@
-
```
Comment (wtrocki) [inline LICENSE:1]: I do not see any change here :D

Also license file missing owner which is wrong :D

File: cmd/cmd.go [hunk 2]
```diff
@@ -1,72 +1,72 @@
...
+	"github.com/spf13/viper"
+)
+
+var cfgFile string
+
+// rootCmd represents the base command when called without any subcommands
+var rootCmd = &cobra.Command{
+	Use:   "charmil",
+	Short: "A brief description of your application",
+	Long: `A longer description that spans multiple lines and likely contains
```
Comment (wtrocki) [inline cmd/cmd.go:20]: Let's keep things refined later. Better to have empty string rather than giberish.

File: LICENSE (hunk 1 above)
Comment (ankithans) [reply LICENSE:1]: done

File: cmd/cmd.go (hunk 2 above)
Comment (ankithans) [reply cmd/cmd.go:20]: done

File: cmd/extensions/cmd.go [hunk 3]
```diff
@@ -1,39 +1,39 @@
...
+	"github.com/aerogear/charmil/cmd/extensions/installed"
+	"github.com/aerogear/charmil/cmd/extensions/list"
+	"github.com/aerogear/charmil/cmd/extensions/remove"
+	"github.com/spf13/cobra"
+)
+
+// ExtensionsCmd represents the extensions command
+var ExtensionsCmd = &cobra.Command{
+	Use:   "extensions",
+	Short: "A brief description of your command",
```
Comment (wtrocki) [inline cmd/extensions/cmd.go]: :) see above ^^

Comment (ankithans) [reply cmd/extensions/cmd.go]: done

File: cmd/extensions/cmd.go [hunk 4]
```diff
@@ -1,39 +1,39 @@
...
+Cobra is a CLI library for Go that empowers applications.
+This application is a tool to generate the needed files
+to quickly create a Cobra application.`,
+	Run: func(cmd *cobra.Command, args []string) {
+		fmt.Println("extensions called")
+	},
+}
+
+func init() {
+	ExtensionsCmd.AddCommand(install.InstallCmd, installed.InstalledCmd, list.ListCmd, remove.RemoveCmd)
```
Comment (wtrocki) [inline cmd/extensions/cmd.go:23]: I think this is good but imagine scenario that:


Root command defines some of the types or factories (top level API) that extensions can consume. 

Those would be for logging, localizer etc. Example:
https://github.com/redhat-developer/app-services-cli/blob/main/pkg/cmd/factory/factory.go#L13-L24

So to realize that architecture we need:

- plugins-core (provides common API for host and extensions to use)
- plugins extension (provides example implementation of extension that uses structure passed from host)
- Host itself that imports extensions into go mod and have some example how plugins core structures are initialized.

File: cmd/extensions/installed/cmd.go [hunk 5]
```diff
@@ -1,34 +1,34 @@
...
+Cobra is a CLI library for Go that empowers applications.
+This application is a tool to generate the needed files
+to quickly create a Cobra application.`,
+	Run: func(cmd *cobra.Command, args []string) {
+		fmt.Println("installed called")
+	},
+}
+
+func init() {
+	// Here you will define your flags and configuration settings.
```
Comment (wtrocki) [inline cmd/extensions/installed/cmd.go]: No comments please. Let's be vigilant to not push any half backed code. 
This is WIP but flagging it early.

**Comment by wtrocki** [comment]
This is good start confirming that we have great understanding of basics, but we are at 10% of the POC progress.

**Comment by wtrocki** [comment]
Btw.. No need to instantly address all the comments. Take it slow and easy. 
Let me know if you have questions about:

```
So to realize that architecture we need:

plugins-core (provides common API for host and extensions to use)
plugins extension (provides example implementation of extension that uses structure passed from host)
Host itself that imports extensions into go mod and have some example how plugins core structures are initialized.

**Comment by ankithans** [comment]
According to my understanding:
plugin core will be like plugin_loader and plugin structure
plugin extension is embeding multiple CLI's using plugin core
> Host itself that imports extensions into go mod and have some example how plugins core structures are initialized. 

Can you explain this more?

**Comment by wtrocki** [comment]
Yes. Core needs to have structure definitions that will be shared. 
What we have right now is 2 packages one depending on each other.. There is simply nothing here to make charmil bring value, but once we introduce core we can bring an opinionated layer of tools that others might use - config abstractions etc.

Look again how this factory works :)

https://github.com/redhat-developer/app-services-cli/blob/main/pkg/cmd/factory/factory.go#L13-L24

**Comment by ankithans** [comment]
okay, now I got the idea about factory. Thanks 👍
working on this PR for now

**Comment by ankithans** [comment]
@wtrocki can we merge this for now? I want to work on present issues on the top of it one by one.

**Comment by wtrocki** [comment]
I personally do not see value of merging yet. Lets work on some feature branch for a while.

We do not want half backed code in main and POC can be done in feature branch - poc or something like this. PRs can still be fast and done against poc branch instead of main. 

There is nothing that have functional factor or significanlty shows the vision or architecture yet

**Comment by wtrocki** [comment]
Little bit of background. Based on my experience in the past fast merges to main branch gived students some form of validation for the work, but we never actually had anything functional in those and ended with looooots of time wasted to cleanup things..

Main alrady has some questionable elements.
Lets work a way in small PRs but keep main clean.


We need to explore what is charmil going to offer for users and it is guaranteed that first approach might not be the right etc.

**Comment by ankithans** [comment]
> fast merges to main branch gived students some form of validation for the work, but we never actually had anything functional

Yup, I can feel this now. Same was going to happen. 
We will work on ```poc``` branch till we have something valuable and refined architecture.

**Pull request closed without merging**